rfcs/0001-shabka-usage-design.md

```diff
@@ -0,0 +1,191 @@
+---
+feature: shabka-usage-design
+start-date: 2019-10-14
+author: Marc 'risson' Schmitt
+co-authors:
+  - Wael 'kalbasit' Nasreddine
+related-issues:
+  - https://github.com/kalbasit/shabka/projects/10
+  - https://github.com/kalbasit/shabka/issues/263
+  - https://github.com/kalbasit/shabka/issues/246
+  - https://github.com/kalbasit/shabka/issues/55
+  - https://github.com/kalbasit/shabka/issues/281
+---
+
+# Summary
+[summary]: #summary
+
+The goal is to design a system where `shabka` is just a tool like
+`home-manager`, and relies on configuration files provided by the user.
+The tools and ways to configure at the disposition of the user would be those:
+
+* the command `shabka`, installed like `home-manager` which basically does the
+  stuff the user wants to do on a regular basis that we currently have in
+  `./scripts` ;
+* a `dotshabka` like directory for the user's hosts and personal preferences
+  settings and default values for `shabka` options, or overriding stuff we
+  define in `shabka`. We will refer to this as `dotshabka`.
+
+# Motivation
+[motivation]: #motivation
+
+This is important in the process of making `shabka` usable for everyone. A
+simple CLI, plus good documentation on how to create configuration files, will
+allow the user to quickly get a hang on the tool.
+
+The planned use cases are very simple for now, and can be described as this:
+
+* Usage on a NixOS system, with root access.
+  - Management of several hosts' configuration
+  - Management of several users per-host.
+  - Management of a single users's configuration hosts-wide.
+  - Management of a single users' configuration per-host.
+* Usage on a Darwin system, with root access.
+  - Management of several hosts' configuration
+  - Management of several users per-host.
+  - Management of a single users's configuration hosts-wide.
+  - Management of a single users' configuration per-host.
+
+The expected outcome is a tool well-documented, easy to use, that can fit easily
+inside an existing workflow of managing NixOS and home-manager configurations.
+The existing configurations would of course need to be tweaked to fit shabka's
+workflow.
+
+# Detailed design
+[design]: #detailed-design
+
+We'll discuss the three aspects previously introduced in [Summary](#summary).
+
+## `shabka` CLI
+
+The `shabka` CLI will be used by the user to do what they can currently do using
+the tools in the `scripts/` directory. This will allow them to do everything they
+could do using the `nixos-rebuild` command, plus building a host, checking what
+changed in their host's generation between two git refs, and maybe even pushing
+their hosts' configurations to Cachix.
+
+Here is what the CLI usage might look like:
+
+### `shabka rebuild`
+
+This will do what `scripts/nixos-rebuild.sh` currently does. Its usage is:
```

Comment:
- For NixOS, this will be `scripts/nixos-rebuild.sh`.
- For Darwin, this will be `scripts/darwin-rebuild.sh`.
- For Linux, non-nixos or for nixos non-root this should probably be around `home-manager`.

I'm thinking that we need to provide clear sub-commands instead of saying this will wrap X or Y tool. We could list them as `build`, `test`, `boot` and `switch`. We may later propose adding `build-vm` and `build-vm-with-bootloader`.

Reply:
Linux (non-nixos) are the subject of another RFC. Adding the rest in the RFC

Reply:
Actually, we could already support `build-vm` and `build-vm-with-bootloader`